### PR TITLE
Move password strings to []byte for active memory clearing

### PR DESCRIPTION
**Pull request by paulmey**
A stab at fixing #3. Please comment...

### CODE REVIEW

File: bmp-string.go [hunk 1]
```diff
@@ -2,9 +2,10 @@ package pkcs12
 
 import (
 	"unicode/utf16"
+	"unicode/utf8"
 )
 
-func bmpString(s string) ([]byte, error) {
+func bmpString(utf8String []byte) ([]byte, error) {
```
Comment (ahmetalpbalkan) [inline bmp-string.go]: Why not []rune

**Comment by paulmey** [comment]
changed []byte -> []rune

**Comment by ahmetalpbalkan** [comment]
http://golang.org/pkg/crypto/x509/#DecryptPEMBlock this guy uses `[]byte`. I don't know what's the deal with between conversion from utf-8 to utf-16 in this context? But `[]rune` sounds fine enough to me.

**Comment by paulmey** [comment]
reverted back to `[]byte` after discussion in #3

File: pkcs12.go [hunk 2]
```diff
@@ -73,9 +73,13 @@ type UnsupportedFormat string
...
-	password = ""
+func ConvertToPEM(pfxData []byte, utf8Password []byte) (blocks []*pem.Block, err error) {
+	p, err := bmpString(utf8Password)
+
+	for i := 0; i < len(utf8Password); i++ {
+		utf8Password[i] = 0
+	}
+
 	if err != nil {
 		return
```
Comment (nathany) [inline pkcs12.go]: The error from bmpString will say "string contains characters...".

Should this code amend the error to indicate that it's the password causing an error?

File: bmp-string.go [hunk 3]
```diff
@@ -12,8 +13,12 @@ func bmpString(s string) ([]byte, error) {
 	//	  EncodeRune returns 0xfffd if the rune does not need special encoding
 	//  - the above RFC provides the info that BMPStrings are NULL terminated.
 
-	rv := make([]byte, 0, 2*len(s)+2)
-	for _, c := range []rune(s) {
+	rv := make([]byte, 0, 2*len(utf8String)+2)
```
Comment (nathany) [inline bmp-string.go]: `utf8.RuneCount` would be more accurate than `len(utf8String)`

(on the other hand, it calls DecodeRun, so likely slower)

File: bmp-string.go (hunk 1 above)
Comment (nathany) [inline bmp-string.go]: One advantage of []rune is that the type system ensures the caller to ConvertToPEM doesn't mix up pfxData and utf8Password (`pfxData []byte, utf8Password []rune`).

But it probably makes sense to stick with what DecryptPEMBlock does.

File: pkcs12.go (hunk 2 above)
Comment (paulmey) [inline pkcs12.go]: :+1:  I like that idea

File: bmp-string.go (hunk 3 above)
Comment (paulmey) [inline bmp-string.go]: I'm ok with wasting a little here. The objective was to allocate at least enough to decourage any reallocation and potential copy of partial passwords. I know we don't have control over what the runtime does, but I can still try to give it hints... :smile:

Comment (nathany) [inline bmp-string.go]: :smiley:

**Comment by paulmey** [comment]
@nathany The error conversion will probably break when you fix the lint issues. Are you ok with taking that in there? If so, I'll merge this.

**Comment by nathany** [comment]
@paulmey Sure, I can fix the conflicts on my (much simpler) branch after.

File: pkcs12.go [hunk 4]
```diff
@@ -73,10 +73,15 @@ type UnsupportedFormat string
 func (e UnsupportedFormat) Error() string { return string(e) }
 
 // ConvertToPEM converts all "safe bags" contained in pfxData to PEM blocks.
-func ConvertToPEM(pfxData []byte, password string) (blocks []*pem.Block, err error) {
-	p, err := bmpString(password)
-	password = ""
+func ConvertToPEM(pfxData []byte, utf8Password []byte) (blocks []*pem.Block, err error) {
```
Comment (nathany) [inline pkcs12.go]: :lipstick: This could also be written `pfxData, utf8Password []byte`.
Dito below.

**Comment by nathany** [comment]
LGTM

File: pkcs12.go (hunk 4 above)
Comment (ahmetalpbalkan) [inline pkcs12.go]: We also don't need to say this is UTF-8

Comment (paulmey) [inline pkcs12.go]: `utf8.DecodeRune` will fail if the underlying encoding is anything else than utf8. Indicating that the byte slice contains uft8 seems useful.

**Pull request merged**